covid19dp_submission/steps/download_analysis.py

```diff
@@ -0,0 +1,127 @@
+
+
+def download_analysis(project, num_analysis, processed_analysis_file, download_target_dir):
+    total_analysis = total_analysis_in_project(project)
+    logger.info(f"total analysis in project {project}: {total_analysis}")
+
+    analysis_array = get_analysis_to_process(project, num_analysis, total_analysis, processed_analysis_file)
+    logger.info(f"number of analysis to process: {len(analysis_array)}")
+
+    download_files(analysis_array, download_target_dir, processed_analysis_file)
+
+
+def total_analysis_in_project(project):
+    count_url = f"https://www.ebi.ac.uk/ena/portal/api/filereportcount?accession={project}&result=analysis"
+    response = requests.get(count_url)
+    if response.status_code != 200:
+        logger.error(f"Error fetching total analysis count for project {project}")
+        response.raise_for_status()
+    return response.json()
+
+
+def get_analysis_to_process(project, num_analysis, total_analysis, processed_analysis_file):
+    offset = 0
+    limit = num_analysis
+    processed_analysis = get_processed_analysis(processed_analysis_file)
+    analysis_for_processing = []
+    while offset < total_analysis:
+        analysis_from_ena = get_analysis_from_ena(project, offset, limit)
+        unprocessed_analysis = filter_out_processed_analysis(analysis_from_ena, processed_analysis)
+
+        if len(analysis_for_processing) + len(unprocessed_analysis) >= num_analysis:
+            analysis_for_processing = analysis_for_processing + unprocessed_analysis[
+                                                                :(num_analysis - len(analysis_for_processing))]
+            break
+        else:
+            analysis_for_processing = analysis_for_processing + unprocessed_analysis
+            offset = offset + limit
+
+    return analysis_for_processing
+
+
+def get_analysis_from_ena(project, offset, limit):
+    analysis_url = f"https://www.ebi.ac.uk/ena/portal/api/filereport?result=analysis&accession={project}&offset={offset}" \
+                   f"&limit={limit}&format=json&fields=run_ref,analysis_accession,submitted_ftp"
+    response = requests.get(analysis_url)
+    if response.status_code != 200:
+        logger.error(f"Error fetching analysis info from ENA for {project}")
+        response.raise_for_status()
+    return response.json()
+
+
+def filter_out_processed_analysis(analysis_array, processed_analysis):
+    unprocessed_analysis = []
+    for analysis in analysis_array:
+        if analysis['analysis_accession'] not in processed_analysis:
+            unprocessed_analysis.append(analysis)
+
+    return unprocessed_analysis
+
+
+def get_processed_analysis(processed_analysis_file):
+    processed_analysis = set()
+    with open(processed_analysis_file, 'r') as file:
+        for line in file:
+            processed_analysis.add(line.split(",")[0])
+    return processed_analysis
+
+
+def download_files(analysis_array, download_target_dir, processed_analysis_file):
```

Comment:
Suggest annotating with retry and jitter as suggested in the stand-up meeting today...

Reply:
added.